lib/records.js

```diff
@@ -0,0 +1,117 @@
+   *
+   * @param {String} key
+   * @param {Function} parser
+   * @returns {Promise} resolves to a Record when complete
+   */
+  async function fetchRecord(key, parser) {
+    const record = await mc.getAsync(key).then(parser, parser)
+    record.key = key
+    return record
+  }
+
+  /**
+   * Fetch a set of records
+   *
+   * @param {Object} config
+   *  @param {String} [object.ip] ip address to fetch
+   *  @param {String} [object.email] email address to fetch
+   *  @param {String} [object.phoneNumber] phone number to fetch
+   *  @param {String} [object.uid] uid to fetch
+   * @returns {Promise} resolves to an object with the following keys:
+   *  `ipRecord`, `reputation`, `emailRecord`, `ipEmailRecord`, `smsRecord`, and `uidRecord`
+   */
+  function fetchRecords(config) {
+    const promises = []
+    const records = {}
+
+    const { ip, email, phoneNumber, uid } = config
+
+    if (ip) {
+      promises.push(
+        // get records and ignore errors by returning a new record
+        fetchRecord(ip, IpRecord.parse).then(ipRecord => records.ipRecord = ipRecord),
+        Promise.resolve(reputationService.get(ip)).then(reputation => records.reputation = reputation)
+      )
+    }
+
+    // The /checkIpOnly endpoint has no email (or phoneNumber)
+    if (email) {
+      promises.push(fetchRecord(email, EmailRecord.parse).then(emailRecord => records.emailRecord = emailRecord))
+    }
+
+    if (ip && email) {
+      promises.push(fetchRecord(ip + email, IpEmailRecord.parse).then(ipEmailRecord => records.ipEmailRecord = ipEmailRecord))
+    }
+
+    // Check against SMS records to make sure that this request can send to this phone number
+    if (phoneNumber) {
+      promises.push(fetchRecord(phoneNumber, smsRecord.parse).then(smsRecord => records.smsRecord = smsRecord))
+    }
+
+    if (uid) {
+      promises.push(fetchRecord(uid, UidRecord.parse).then(uidRecord => records.uidRecord = uidRecord))
+    }
+
+    return Promise.all(promises).then(() => records)
+  }
+
+  /**
+   * Save a record
+   *
+   * @param {Record} record
+   * @returns
+   */
+  function setRecord(record) {
+    const lifetime = Math.max(recordLifetimeSeconds, record.getMinLifetimeMS() / 1000)
+    return mc.setAsync(record.key, marshallRecordForStorage(record), lifetime)
+  }
+
+  /**
+   * Marshall a record for persistent storage
+   *
+   * @param {Record} record
+   * @returns {Object}
+   */
+  function marshallRecordForStorage (record) {
+    const marshalled = {}
+
+    for (const key of Object.keys(record)) {
+      if (key !== 'key' && typeof record[key] !== 'function') {
+        marshalled[key] = record[key]
+      }
+    }
+
+    return marshalled
+  }
+
+  /**
+   * Save records
+   *
+   * @param {Record[]} records to save
+   * @returns {Promise} Resolves when complete
+   */
+  function setRecords(...records) {
+    const promises = records.map(record => record && setRecord(record))
```

Comment:
The `record && setRecord(record)` part seem strange to me, when would a record be undefined and have called `setRecords`?